workers/unity/Packages/com.improbable.gdk.deploymentlauncher/TaskManager.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Collections.Generic;
+using Improbable.Gdk.DeploymentManager.Commands;
+using UnityEditor;
+
+namespace Improbable.Gdk.DeploymentManager
+{
+    internal class TaskManager
+    {
+        public enum QueueMode
+        {
+            Normal,
+            ForceNext
+        }
+
+        public bool IsActive => CurrentTask != null;
+        public IWrappedTask CurrentTask { get; private set; }
+        public readonly List<IWrappedTask> CompletedTasks = new List<IWrappedTask>();
+
+        private List<Func<IWrappedTask>> queuedTasks = new List<Func<IWrappedTask>>();
+        private bool isLocked;
+        private Func<IWrappedTask> queuedAuthTask;
+
+        public void ClearResults()
+        {
+            CompletedTasks.Clear();
+        }
+
+        public void Cancel()
+        {
+            CurrentTask?.Cancel();
+            queuedTasks.Clear();
+        }
+
+        public void Upload(AssemblyConfig config, QueueMode mode = QueueMode.Normal)
+        {
+            AddTask(mode, () => Assembly.UploadAsync(config));
+        }
+
+        public void Launch(string projectName, string assemblyName, BaseDeploymentConfig config,
+            QueueMode mode = QueueMode.Normal)
+        {
+            AddTask(mode, () => Deployment.LaunchAsync(projectName, assemblyName, config));
+        }
+
+        public void List(string projectName, QueueMode mode = QueueMode.Normal)
+        {
+            AddTask(mode, () => Deployment.ListAsync(projectName));
+        }
+
+        public void Stop(DeploymentInfo info, QueueMode mode = QueueMode.Normal)
+        {
+            AddTask(mode, () => Deployment.StopAsync(info));
+        }
+
+        public void Auth()
+        {
+            queuedAuthTask = Authentication.Authenticate;
+        }
+
+        public void Update()
+        {
+            if (!IsActive)
+            {
+                // Always prefer the auth task over any queued t asks.
+                if (queuedAuthTask != null)
+                {
+                    CurrentTask = queuedAuthTask();
+                    queuedAuthTask = null;
+
+                    if (!isLocked)
```

Comment:
Can you simplify this a bit by moving the lock check out, and change the last if/else to
```if(CurrentTask != null && !isLocked)
{}
else if (isLocked)
{}```?

Reply:
I can introduce a local bool and so something like: 

```
            if (!IsActive)
            {
                var hasStartedTask = false;

                // Always prefer the auth task over any queued t asks.
                if (queuedAuthTask != null)
                {
                    CurrentTask = queuedAuthTask();
                    queuedAuthTask = null;
                    hasStartedTask = true;
                }
                else if (queuedTasks.Count > 0)
                {
                    var queuedTask = queuedTasks[0];
                    queuedTasks.RemoveAt(0);

                    CurrentTask = queuedTask();
                    hasStartedTask = true;
                }
                
                
                if (hasStartedTask)
                {
                    if (!isLocked)
                    {
                        EditorApplication.LockReloadAssemblies();
                        isLocked = true;
                    }
                }
                else if (isLocked)
                {
                    EditorApplication.UnlockReloadAssemblies();
                    isLocked = false;
                }
            }
```

The main problem is we cannot enter the `else if (isLocked)` block if we have started a task this same `Update()` call, otherwise it will never actually lock the assemblies.